### PR TITLE
Support uploading oci-image resources.

### PR DESCRIPTION
**Pull request by facundobatista**

### CODE REVIEW

**Review by cjp256 — APPROVED** [review]

File: charmcraft/commands/store/__init__.py [hunk 1]
```diff
@@ -1132,8 +1172,13 @@ class UploadResourceCommand(BaseCommand):
         specified charm. This charm needs to have the resource declared
         in its metadata (in a preoviously uploaded to Charmhub revision).
 
-        to the packaging standard. This command will finish successfully
-        once the package is approved by Charmhub.
+        The resource can be a file from your computer (use the '--filepath'
+        option) or an OCI Image (use the '--image' option).
+
+        The OCI image description uses the [organization/]name[:reference]
```
Comment (cmatsuoka) [inline charmcraft/commands/store/__init__.py:1178]: The `[:reference]` format description here applies only to `[:tag]`, shouldn't we also mention `[@digest]` here to be consistent with the `--image` parameter help message below?

**Review by cmatsuoka — APPROVED** [review]
Looks good, just added a note about documentation consistency.

File: charmcraft/commands/store/__init__.py [hunk 2]
```diff
@@ -1120,6 +1124,42 @@ def run(self, parsed_args):
...
+    if '/' in value:
+        if value.count('/') > 1:
+            raise _BadOCIImageSpecError(
+                "The registry server cannot be specified as part of the image")
+        orga, value = value.split('/')
+    else:
+        orga = 'library'
+
+    # get the digest XOR tag
+    if '@' in value:
```
[the remark this answers is not in the document]
Comment (cmatsuoka) [reply charmcraft/commands/store/__init__.py:1146]: Agree with @cjp256, handling the error case just once is better.

File: tests/test_help.py [hunk 3]
```diff
@@ -49,7 +49,12 @@ def test_aesthetic_args_options_msg(command, config):
     class FakeParser:
         """A fake to get the arguments added."""
 
+        def add_mutually_exclusive_group(self, *args, **kwargs):
+            """Return self, as it be used to add arguments too."""
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply tests/test_help.py:53]: ja, that came from a "would be" :p

File: tests/test_help.py [hunk 4]
```diff
@@ -49,7 +49,12 @@ def test_aesthetic_args_options_msg(command, config):
...
+            return self
+
         def add_argument(self, *args, **kwargs):
+            """Verify that all commands must have a correctly formatted help."""
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply tests/test_help.py:57]: thanks

File: charmcraft/commands/store/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (facundobatista) [reply charmcraft/commands/store/__init__.py:1146]: Nice! Improved.

File: charmcraft/commands/store/__init__.py (hunk 1 above)
Comment (facundobatista) [reply charmcraft/commands/store/__init__.py:1178]: yes! thanks, good catch! improved

**Pull request merged**